bin/parse-legacy-file

```diff
@@ -12,21 +14,243 @@ is_strict_equality_version_constraint() {
+        return 0
+      fi
+    ;;
+
+  ">=")
+      if ! is_stable "$current_rest"; then
+        return 1
+      fi
+
+      if is_version_eq "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+
+      if is_version_gt "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+    ;;
+
+  "<")
+      if ! is_stable "$current_rest"; then
+        return 1
+      fi
+
+      if is_version_eq "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 1
+      fi
+
+      if ! is_version_gt "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+    ;;
+
+  "<=")
+      if ! is_stable "$current_rest"; then
+        return 1
+      fi
+
+      if is_version_eq "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+
+      if ! is_version_gt "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+    ;;
+
+  "~>")
+      if ! is_stable "$current_rest"; then
+        return 1
+      fi
+
+      if ! is_version_major_minor_eq "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 1
+      fi
+
+      if is_version_eq "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+
+      if is_version_gt "$allowed_version" "$allowed_minor" "$allowed_patch" "$allowed_rest" "$current_major" "$current_minor" "$current_patch" "$current_rest"; then
+        return 0
+      fi
+    ;;
+
+  esac
+
+  return 1
+}
 
 read_version_from_main_tf() {
   local -r version_file="$1"
-  local -r required_version_constraint="$(grep --no-filename \
-    --recursive \
-    --fixed-strings \
-    required_version \
-    "${version_file}" |
-    sed -E 's/[[:space:]]*required_version[[:space:]]*=//' |
-    tr -d '" ')"
+  local -r required_version_constraint="$(awk -F\" '/^\s*required_version\b*/{print $2}' "${version_file}")"
 
   [[ -z ${required_version_constraint} ]] && return 0
 
-  if is_strict_equality_version_constraint "${required_version_constraint}"; then
-    tr -d '=' <<<"${required_version_constraint}"
+  # todo doing this on every dir change is expensive. What is a good alternative?
+  local all_versions
+  IFS=" " read -r -a all_versions <<<"$("${__dirname}/list-all")"
```

Comment:
Not really sure what the best approach is here. This is pretty expensive to do every single time since it's a network request. Also, in case there is no network, this will most definitely fail.

Perhaps what should be done here is only check local versions if offline (though, how do we check if a user is offline)? And when they are online, perhaps, we can cache the results when `list-all` is run.

Any thoughts?